icu4c/source/common/locid.cpp

```diff
@@ -424,56 +424,47 @@ Locale::Locale(const Locale &other)
     *this = other;
 }
 
-Locale &Locale::operator=(const Locale &other)
-{
+Locale& Locale::operator=(const Locale& other) {
     if (this == &other) {
         return *this;
     }
 
-    /* Free our current storage */
+    if (fullName != fullNameBuffer) {
+        uprv_free(fullName);
+    }
+
     if (baseName != fullName) {
         uprv_free(baseName);
     }
-    baseName = NULL;
-    if(fullName != fullNameBuffer) {
-        uprv_free(fullName);
-        fullName = fullNameBuffer;
-    }
 
-    /* Allocate the full name if necessary */
-    if(other.fullName != other.fullNameBuffer) {
-        fullName = (char *)uprv_malloc(sizeof(char)*(uprv_strlen(other.fullName)+1));
-        if (fullName == NULL) {
-            // if memory allocation fails, set this object to bogus.
-            fIsBogus = TRUE;
-            return *this;
-        }
+    baseName = fullName = fullNameBuffer;
+    fIsBogus = TRUE;
+
+    if (other.fullName == nullptr) {
+      fullName = nullptr;
+    } else if (other.fullName == other.fullNameBuffer) {
+      uprv_strcpy(fullNameBuffer, other.fullNameBuffer);
+    } else {
+      fullName = uprv_strdup(other.fullName);
```

Comment:
If `uprv_strdup` fails to allocate, then the locale isn't set to bogus anymore.

Reply:
I don't think I understand what you mean, could you please elaborate?

Reply:
Ah, never mind. You set `fIsBogus` to true ahead of time and only change it afterwords.